### PR TITLE
Update chaincode APIs

### LINKED ISSUE #875: we need the shim to provide the transaction timestamp to the contract
In our domain, timestamps at the nano level are critical to sorting out the history of events relating to an asset. This in fact is equally important in any domain, even finance. You must know the time at which something was signed.

Currently, our contract generates timestamps when the event does not carry one, but on today's interlock call with @mastersingh24, we confirmed that we will _never_ achieve consensus in a network with actual consensus running because the state can not converge on one timestamp unless it comes in in the transaction. We absolutely cannot assume that we can force every application and device in the entire world to send timestamps, so we are suddenly very stuck.

After some discussion of messy front ends to inject timestamps, Gari suggested that the transaction already has a timestamp that is generated at the moment of initial creation and then forever propagated, and that contracts should be asking the shim for that timestamp in order to have something to define the time accurately. Since the transaction is propagated, every peer will get the same txn time if this is implemented.

This issue is as high a priority as we can imagine. We cannot run on pbft with timestamps generated inside the contract, but we must have timestamps. 

Since the transaction UUID is already available in the stub:

// ChaincodeStub for shim side handling.
type ChaincodeStub struct {
    UUID string
}

May we please get the transaction timestamp in there as well? 

Edit: Note that the timestamp we need is defined in [3.1.2.1 Transaction Data Structure](https://github.com/hyperledger/fabric/blob/master/docs/protocol-spec.md#3121-transaction-data-structure).

### PR DESCRIPTION
**Pull request by binhn**
Passing transaction timestamp to the chaincode context and update the APIs
## Description

Chaincode needs the transaction creation timestamp. The fabric includes the timestamp in the context passed to the chaincode on init, invoke, and query calls. This change also includes update to the chaincode API document and godoc comments. Unit-test for the timestamp is added to the chaincode_example01, which is run as part of the test-suite. 
## Motivation and Context

Fixes #875
## How Has This Been Tested?

Run the unit-tests or the chaincode_example01
## Checklist:

- [x ] I have added a [Signed-off-by](https://github.com/hyperledger/fabric/blob/master/CONTRIBUTING.md#legal-stuff)
- [x ] Either no new documentation is required by this change, OR I added new documentation
- [ x] Either no new tests are required by this change, OR I added new tests
- [x ] I have run [goimports](https://godoc.org/golang.org/x/tools/cmd/goimports), [go vet](https://golang.org/cmd/vet/), and [golint](https://github.com/golang/lint). I have cleaned up all valid errors and warnings in code I have added or modified. These tools may generate false positives. Don't be worried about ignoring some errors or warnings. The goal is clean, consistent, and readable code.

Signed-off-by: binhn@us.ibm.com

### CODE REVIEW

File: core/chaincode/shim/chaincode.go [hunk 1]
```diff
@@ -252,39 +264,39 @@ func (stub *ChaincodeStub) init(uuid string, secContext *pb.ChaincodeSecurityCon
 
 // ------------- Call Chaincode functions ---------------
 
-// InvokeChaincode function can be invoked by a chaincode to execute another chaincode.
+// InvokeChaincode locally calls the specified chaincode `Invoke`.
```
Comment (srderson) [inline core/chaincode/shim/chaincode.go]: People always ask me if this creates a second transaction. Should we state that it does not in the comment? Maybe that is what you were trying to imply by locally?

File: core/chaincode/shim/chaincode.go [hunk 2]
```diff
@@ -252,39 +264,39 @@ func (stub *ChaincodeStub) init(uuid string, secContext *pb.ChaincodeSecurityCon
...
 func (stub *ChaincodeStub) InvokeChaincode(chaincodeName string, function string, args []string) ([]byte, error) {
 	return handler.handleInvokeChaincode(chaincodeName, function, args, stub.UUID)
 }
 
-// QueryChaincode function can be invoked by a chaincode to query another chaincode.
+// QueryChaincode locally calls the specified chaincode `Query`.
```
Comment (srderson) [inline core/chaincode/shim/chaincode.go]: People always ask me if this creates a second transaction. Should we state that it does not in the comment?

**Comment by binhn** [comment]
@srderson yes, that's what I meant by "locally", but i agreed what it's worth more detail explanation here. I'll provide an update

**Comment by binhn** [comment]
@srderson i rebased and force push to update the pr.  Please merge

**Comment by srderson** [comment]
LGTM

**Pull request merged**